consensus/istanbul/validator/weighted.go

```diff
@@ -586,41 +586,89 @@ func (valSet *weightedCouncil) Refresh(hash common.Hash, blockNum uint64, stakin
+	numValidators := len(valSet.validators)
 	weightedValidators := make([]*weightedValidator, numValidators)
 	stakingAmounts := make([]float64, numValidators)
-	totalStaking := float64(0)
+	addedStaking := make([]bool, len(stakingInfo.CouncilNodeAddrs))
+
 	for vIdx, val := range valSet.validators {
 		weightedVal, ok := val.(*weightedValidator)
 		if !ok {
-			return errors.New(fmt.Sprintf("not weightedValidator. val=%s", val.Address().String()))
+			return nil, nil, errors.New(fmt.Sprintf("not weightedValidator. val=%s", val.Address().String()))
 		}
 		weightedValidators[vIdx] = weightedVal
 
-		i, err := newStakingInfo.GetIndexByNodeId(weightedVal.address)
+		sIdx, err := stakingInfo.GetIndexByNodeId(weightedVal.address)
 		if err == nil {
-			weightedVal.SetRewardAddress(newStakingInfo.CouncilRewardAddrs[i])
-			tempStakingAmount := float64(newStakingInfo.CouncilStakingAmounts[i])
-			if newStakingInfo.UseGini {
-				tempStakingAmount = math.Round(math.Pow(tempStakingAmount, 1.0/(1+newStakingInfo.Gini)))
-			}
-			stakingAmounts[vIdx] = tempStakingAmount
-			totalStaking += tempStakingAmount
+			rewardAddr := stakingInfo.CouncilRewardAddrs[sIdx]
+			weightedVal.SetRewardAddress(rewardAddr)
+			stakingAmounts[vIdx] = float64(stakingInfo.CouncilStakingAmounts[sIdx])
+			addedStaking[sIdx] = true
 		} else {
 			weightedVal.SetRewardAddress(common.Address{})
 		}
 	}
 
-	// Update each validator's weight based on the ratio of its staking amount vs. the total staking amount.
+	for sIdx, isAdded := range addedStaking {
+		if isAdded {
+			continue
+		}
+		for vIdx, val := range weightedValidators {
+			if val.RewardAddress() == stakingInfo.CouncilRewardAddrs[sIdx] {
+				stakingAmounts[vIdx] += float64(stakingInfo.CouncilStakingAmounts[sIdx])
+				break
+			}
+		}
+	}
+
+	logger.Debug("stakingAmounts of validators", "validators", weightedValidators, "stakingAmounts", stakingAmounts)
+	return weightedValidators, stakingAmounts, nil
+}
+
+// calculate totalAmount of stakingAmounts
+// if UseGini is true, gini is reflected to stakingAmounts
+func calcTotalAmount(stakingInfo *reward.StakingInfo, stakingAmounts []float64) float64 {
+	totalStaking := float64(0)
+	if stakingInfo.UseGini && len(stakingInfo.CouncilNodeAddrs) != 0 {
+		stakingInfo.Gini = reward.CalcGiniCoefficient(stakingAmounts)
+
+		for i := range stakingAmounts {
+			stakingAmounts[i] = math.Round(math.Pow(stakingAmounts[i], 1.0/(1+stakingInfo.Gini)))
+			totalStaking += stakingAmounts[i]
+		}
+	} else {
+		for _, stakingAmount := range stakingAmounts {
+			totalStaking += stakingAmount
+		}
+	}
+
+	logger.Debug("calculate totalStaking", "UseGini", stakingInfo.UseGini, "Gini", stakingInfo.Gini, "totalStaking", totalStaking, "stakingAmounts", stakingAmounts)
+	return totalStaking
+}
+
+// Update each validator's weight based on the ratio of its staking amount vs. the total staking amount.
```

Comment:
Minor. But I remember our function comment policy is something like "FUNCTION_NAME verb ...". For this function, it would be "calcWeight updates each validator's..."

Reply:
reflected